src/webui/WebUI.jl

```diff
@@ -0,0 +1,227 @@
+    )
+    branch = sendrecv(zsock, regp; nretry=5)
+    if branch === nothing || get(branch.metadata, "error", nothing) !== nothing
+        if branch === nothing
+            msg = "ERROR: Registrator backend service unreachable"
+        else
+            msg = "Registration failed: " * branch.metadata["error"]
+        end
+        state = :errored
+    else
+        title, body = pull_request_contents(;
+            registration_type=get(branch.metadata, "kind", ""),
+            package=regdata.project.name,
+            repo=web_url(regdata.repo),
+            user=display_user(regdata.user),
+            gitref=regdata.ref,
+            version=regdata.project.version,
+            commit=regdata.commit,
+            patch_notes=regdata.notes,
+        )
+
+        # Make the PR.
+        pr = @gf make_registration_request(REGISTRY[], branch.branch, title, body)
+        if pr === nothing
+            msg = "Registration failed: Making pull request failed"
+            state = :errored
+        else
+            url = web_url(pr)
+            msg = """Registry PR successfully created, see it <a href="$url" target="_blank">here</a>!"""
+            state = :success
+        end
+    end
+    @debug msg
+    REGISTRATIONS[regdata.commit] = RegistrationState(msg, state)
+    nothing
+end
+
+function request_processor(zsock::RequestSocket)
+    do_action() = action(take!(event_queue), zsock)
+    handle_exception(ex) = ex isa InvalidStateException && ex.state === :closed ? :exit : :continue
+    keep_running() = isopen(httpsock[])
+    recover("request_processor", keep_running, do_action, handle_exception)
+end
+
+function start_server(ip::IPAddr, port::Int)
+    httpsock[] = Sockets.listen(ip, port)
+    @app server = (
+        error_handler,
+        pathmatch(ROUTE_INDEX, index),
+        pathmatch(ROUTE_AUTH, auth),
+        pathmatch(ROUTE_CALLBACK, callback),
+        pathmatch(ROUTE_SELECT, select),
+        pathmatch(ROUTE_REGISTER, register),
+        pathmatch(ROUTE_STATUS, status),
+        r -> html(404, "Page not found"),
+    )
+    do_action() = begin; t = serve(server, ip, port; server=httpsock[], readtimeout=0); wait(t); end
+    handle_exception(ex) = ex isa Base.IOError && ex.code == -103 ? :exit : :continue
+    keep_running() = isopen(httpsock[])
+    @info "Serving" ip port
+    recover("webui", keep_running, do_action, handle_exception)
+end
+
+function main(config::AbstractString="config.toml")
+    if !isempty(ARGS) && isfile(ARGS[1])
+        merge!(CONFIG, TOML.parsefile(ARGS[1])["web"])
+    else
+        merge!(CONFIG, TOML.parsefile(config)["web"])
+    end
+
+    global_logger(SimpleLogger(stdout, get_log_level(CONFIG["log_level"])))
+    zsock = RequestSocket()
+
+    if haskey(CONFIG, "route_prefix")
+        for r in [:ROUTE_INDEX, :ROUTE_AUTH, :ROUTE_CALLBACK, :ROUTE_SELECT, :ROUTE_REGISTER, :ROUTE_STATUS]
+            @eval $r = CONFIG["route_prefix"] * $r
```

Comment:
this is kinda gross. Is it required to be this way by MUX somehow?
If not, can we insead have a  global `const ROUTE = Dict(:INDEX=> ....
which is then mutated by this function.